### DISCUSSION TITLE
Improve prefetching w/ ViewTransitions

### DISCUSSION THREAD

**Discussion by kurtextrem** [withastro/roadmap#661] [Proposal]
### Body

# Summary

Currently, the view transition "router" does not cancel previous prefetches and triggers new prefetches on hover, touchstart and focus [immediately](https://github.com/withastro/astro/blob/main/packages/astro/components/ViewTransitions.astro#L335). 
If you think about an eCommerce shop where you have a grid of products on any product listing page, this will lead to many many prefetches getting triggered at once.

# Background & Motivation

Instant.page which innovated the approach of prefetching on hover has a default delay of 65 ms and is [customizable](https://instant.page/intensity#:~:text=Adjusting%20the%20delay%20on%20hover) for this reason. 
Additionally, a previous prefetch is not cancelled when a new one is initiated, which could lead to congestion of the network (think 15 prefetches at once, as there is no limit of active `<link rel="prefetch"` in Chrome), this is more an issue on mobile networks (e.g. flaky internet etc. leading to request stalling or similar).

# Goals

- Add a customizable delay before prefetch is initiated
- Cancel previous prefetch when triggering a new one on mobile
- Possibly set priority of the prefetch (using the `importance` attribute), while we do so, we could also increase the priority of the fetch itself

# Example / Implementation

## Fetch delay
`<ViewTransitions fetchDelayMs={65}>`

- introduce new attribute called `fetchDelayMs`, set to 65 by default
- needs docs update

## Cancel previous prefetch

Instead of checking if a node with the to-be-prefetched URL [exists](https://github.com/withastro/astro/blob/main/packages/astro/components/ViewTransitions.astro#L254), we use one prefetcher node and clear the attributes before triggering a new prefetch on mobile.

In addition, currently, no prefetch is made when we're on [3G or 2G](https://github.com/withastro/astro/blob/main/packages/astro/components/ViewTransitions.astro#L257), however, I'd say we can make a prefetch on 3G and 2G if we cancel the previous one (same for `saveData`). A prefetch on 2G/3G definitely makes sense, because whenever a user touches a link (`touchstart`), on a mobile it's likely it becomes a click afterwards (as there is no real hover mechanism on mobile). So any head start here has immediate effect on loading performance on slow networks.

## Prefetch & Fetch priority

As the [`fetch`](https://github.com/withastro/astro/blob/main/packages/astro/components/ViewTransitions.astro#L62C3) is triggered only when we're transitioning to the next page, it makes sense to run it with high priority:

```diff
-const res = await fetch(href);
+const res = await fetch(href, { priority: 'high' });
```

On desktop, as users could hover over multiple elements, it does not make sense to add `importance` to the `link` tag. However, on mobile, as mentioned before, a touch likely means we'll trigger a navigation, thus it makes sense to add importance=high there.

https://github.com/withastro/astro/blob/main/packages/astro/components/ViewTransitions.astro#L261 `link.setAttribute('importance', 'high');`

**Comment by matthewp** [comment]
Thanks! I think I understand most of this. I don't quite understand the delay though. As you said, on mobile that's likely going to be navigation, so delaying 65ms just slows down the navigation.

But re-using a link node to change the `href` in order to cancel previous hovers makes sense to me for sure. Happy to receive a PR for this enhancement!

**Comment by kurtextrem** [comment]
I'm happy to elaborate on the delay: Whenever a user starts a scroll motion on a smartphone, the `touchstart` event is triggered. Now think of the eCommerce example again, a product listing page might have many items next to each other.
In that scenario any scroll might start on an `<a>` element, which means we'll prefetch -> cancel -> prefetch -> cancel etc., only because the user is scrolling around (via touch). For a desktop it's similar, using the mouse you could end up on many `<a>` tags while you're moving your mouse to a target (e.g. from all the way right to all the way left of the screen).

So some kind of delay ensures it is more likely that an initiated prefetch is actually useful and not only wasted network transfer/CPU time.

Instant.page goes a little bit beyond the `touchstart` delay, as they also add [a check to `mouseover`](https://github.com/instantpage/instant.page/blob/master/instantpage.js#L181) with a delay of `1111`, to make sure any prefetch is not triggered due to (unwanted) touch inputs. However, I'm not entirely sure how `1111` is calculated, which is why I did not mention it here. Tagging @dieulot, maybe he is willing to share some insights :)

**Reply by matthewp** [reply]
Thanks for elaborating @kurtextrem. In that case I think both of these additions are worth pursuing to some extent. Still curious about the exact conditions here. Should these delays be focused on mobile? On the slower networks only?

**Reply by dieulot** [reply]
The 65ms delay is appropriate for desktop only. On mobile, users touch a link (click timestamp minus touchstart timestamp) only 90 ms on average, so there’s unfortunately no time for using a delay as another heuristic.

Regarding the 1111ms last touch check in mouseover, going by memory from a few years ago (!) it’s there because tapping a link ends up triggering regular mouse events afterwards. I observed that on Android (I guess I tested in multiple browsers) it took one second of holding your finger on a link for a contextual menu (with options like “Open in new tab” etc.) to appear, at which point it should mean that the mouseover wasn’t triggered by a touch event (which would have already triggered a prefetch), but is indeed the user hovering their mouse (or equivalent device). I added 100 ms (well, 111) as a safety precaution in case Android/browsers would mess that timing if they were to freeze at that time. (I’d like to eventually revisit the issue and add a comment to that magic number in instant.page’s source.)

**Reply by kurtextrem** [reply]
Thank you for the explanation! Maybe it's also worth exploring `pointer*` events to combine touch & mouse listeners (+ add support for e.g. pens).

**Reply by dieulot** [reply]
I doubt it. Mouse and touch ideally use different prefetching strategies, and in instant.page they’re configurable. The only advantage left, pens, are too much of a minority to worry about for my part. (I don’t handle keyboard navigation either.)

**Comment by martrapp** [comment]
As often, it's a matter of finding a setting that usually works well, even if we can only speculate about the nature of the applications.
For small websites with few links it might be perfect if all links are prefetched as soon as they scroll into view, for the online store example this is of course the killer.
Prefetching makes client-side routing snappier, but it can also be defined independently. connection.saveData isn't very reliable, and not everybody has a unlimited flat rate. Imho, we should prefetch too little rather than too much by default and allow users to use more aggressive strategies as add-ons.

**Reply by kurtextrem** [reply]
> connection.saveData isn't very reliable, and not everybody has a unlimited flat rate.

I certainly agree on that part.

To give a comparison, Next.js will prefetch any link that is scrolled into viewport (or is in viewport) if `<Link prefetch={true}>` is set. Similar is the library Quicklink, which also prerenders/prefetches anything that is scrolled into viewport (customizable amount).
Apart from those libs, there is also the speculationrules spec, where you can set `eagerness` (of prefetching). The value `moderate` matches "prefetching on hover", with a delay of [200 ms](https://github.com/WICG/nav-speculation/blob/main/chrome-2023q1-experiment-overview.md#specifying-eagerness). This is imho a bit too high, but I've contacted Domenic from the Chromium team to find out what lead to 200 ms. Maybe we can use that as a more sensible default if their reason matches our goals (personally I'd say 200 is on the very high side, because TTFB is rarely under 200 ms).

In any case, we should definitely keep the customization option + enrich docs with examples. From my own experience 65 ms does not lead to too many wasted prefetches (and if so, it mostly impacts desktop).

**Reply by matthewp** [reply]
Ah, I like the idea of `eagerness`. I could see this as being a new prop on the `ViewTransitions` component.

I plan on writing a proposal soon that's a sort of "view transition additions" proposal that includes a bunch of smaller things like new events, a JS API for triggering navigation, and likely this could be wrapped up in this as well.

**Comment by martrapp** [comment]
I should probably write a little more clearly: What worries me is the lack of separation of concerns: there is an official prefetch integration (@astrojs/prefetch) and an opinionated strategy within ViewTransitions. Wouldn't it make more sense to inform the user about the need for a prefetch strategy and to keep additional prefetching out of the ViewTransition router?

**Reply by kurtextrem** [reply]
I'm not sure if that isn't mixing two things. In ViewTransitions we're proposing a delay for prefetching on strong user signal (hover), in @astrojs/prefetch it looks like to me it's an alternative to Quicklink/Next.js (prefetch what's in viewport). The latter has a much higher chance for prefetching waste (as not every link in viewport will be clicked, only one of N)

We could of course also say we want to change @astrojs/prefetch... but that might explode the scope a bit? (e.g. for @astrojs/prefetch we could even introduce `prerender` instead of only prefetch, which Quicklink also did and my fork of Instant.page)

**Reply by matthewp** [reply]
I think prefetching is important to ensure snappiness. In Next.js when you use the `<Link>` component it prefetches the code for that page. For Astro, prefetching the page itself is the equivalent. I agree that having the separate prefetch integration makes this a bit confusing, we need to figure out a way to combine then in the future.

In the meantime I think we can give better defaults and more control. @kurtextrem's idea above of `<ViewTransition prefetch="moderate" />` with other options of `'none'` and `'fast'` or whatever. I could also see allowing control on links themselves, `<a href="/other" data-astro-prefetch="none">`
